glslc/test/option_dash_M.py

```diff
@@ -0,0 +1,738 @@
+    """Tests -M with single input file which does #include another file in a
+    subdirectory of current directory and is represented in absolute path.
+    e.g. glslc -M /usr/local/a.vert
+      => a.vert.spv: /usr/local/a.vert /usr/local/include/b.vert
+    """
+    environment = Directory('.', [
+        Directory('include', [File('b.vert', 'void foo(){}\n')]),
+    ])
+    shader_main = FileShader('#version 140\n#include "include/b.vert"',
+                             '.vert')
+    glslc_args = ['-M', shader_main]
+    dependency_rules_expected = [{
+        'target': shader_main,
+        'target_extension': '.spv',
+        'dependency': {shader_main}
+        # The dependency here is not complete.  we can not get the absolute
+        # path of include/b.vert here. It will be added in
+        # check_stdout_dependency_info()
+    }]
+
+    def check_stdout_dependency_info(self, status):
+        # Add the absolute path of include/b.vert to the dependency set
+        self.dependency_rules_expected[0]['dependency'].add(os.path.dirname(
+            self.shader_main.filename) + '/include/b.vert')
+        return DependencyInfoStdoutMatch.check_stdout_dependency_info(self,
+                                                                      status)
+
+
+@inside_glslc_testsuite('OptionsCapM')
+class TestDashCapMOverridesOtherModes(DependencyInfoStdoutMatch):
+    """Tests -M overrides other compiler mode options, includeing -E, -c and -S.
+    """
+    environment = Directory('.', [
+        File('a.vert', MINIMAL_SHADER),
+        File('b.vert', MINIMAL_SHADER),
+    ])
+    glslc_args = ['-M', '-E', '-c', '-S', 'a.vert', 'b.vert']
+    dependency_rules_expected = [{'target': 'a.vert.spv',
+                                  'dependency': {'a.vert'}},
+                                 {'target': 'b.vert.spv',
+                                  'dependency': {'b.vert'}}]
+
+
+@inside_glslc_testsuite('OptionsCapM')
+class TestDashCapMMEquivalentToCapM(DependencyInfoStdoutMatch):
+    """Tests that -MM behaves as -M.
+    e.g. glslc -MM shader.vert
+      => shader.vert.spv: shader.vert
+    """
+    environment = EMPTY_SHADER_IN_CURDIR
+    glslc_args = ['-MM', 'shader.vert']
+    dependency_rules_expected = [{'target': 'shader.vert.spv',
+                                  'dependency': {'shader.vert'}}]
+
+
+@inside_glslc_testsuite('OptionsCapM')
+class TestDashCapMImpliesDashCapE(DependencyInfoStdoutMatch,
+                                  expect.NoOutputOnStderr):
+    """Tests that -M implies -E, a .glsl file without an explict stage should
+    not generate an error.
+    e.g. glslc -M shader.glsl
+      => shader.spv: shader.glsl
+         <no error message should be generated>
+    """
+    environment = Directory('.', [File('shader.glsl', MINIMAL_SHADER)])
+    glslc_args = ['-M', 'shader.glsl']
+    dependency_rules_expected = [{'target': 'shader.spv',
+                                  'dependency': {'shader.glsl'}}]
```

Comment:
Woops.  I was wrong.  You explained it in the comment, and I forgot how the dependency_rules_expected mixin variable worked. (It's been a long day)